Add Least-Square harmonic analysis method (for muiltiband-noise branch)

tuxzz opened the pull request:
## Changes
1. Add Least-Square harmonic analysis method. (qhm.c and qhm.h, need 3rd-party lib, see below)
2. Update libpyin to the latest version.
3. Fix inline functions.
## About QHM-AIR
### Requirements
- CBLAS for zgemm
- LAPACKE for zgels and zgelsd

Tested on OpenBLAS and reference blas/lapack
### Advantage

Obviously quality improving compared to FFT Peak picking.
By looking: http://img.vim-cn.com/4a/2a8b0a135ed0e0fd35ad0327570ba29922a23b.png
By listening: https://mega.nz/#!60MVwTxY!01Vy9k-50-BqpWPqMcUOIUv0iRTTTSj6fRjTa2zLTIk
Voice source: Gakkou Gurashi
### Disadvantage

Slower than fft peak picking. Especially on using zgelsd method for LS.
### Works Cited
- Pantazis Y, Rosec O, Stylianou Y. Iterative estimation of sinusoidal signal parameters[J]. Signal Processing Letters, IEEE, 2010, 17(5): 461-464.
- Degottex G, Stylianou Y. Analysis and synthesis of speech using an adaptive full-band harmonic model[J]. Audio, Speech, and Language Processing, IEEE Transactions on, 2013, 21(10): 2085-2095.

**I hope future version of morsampler can have QHM-AIR as an optional harmonic analysis method.**

tuxzz committed:
update pyin

tuxzz committed:
update pyin

tuxzz committed:
add qhm
add works cited
fix inline functions

tuxzz committed:
fix redefined macro

tuxzz committed:
using ?gels by default for better speed

tuxzz committed:
fix for reference blas/lapack

tuxzz committed:
handle invalid parameter

tuxzz committed:
optional disable progress output

tuxzz committed:
add requirement

tuxzz committed:
documented, clean unused variable

tuxzz committed:
use qhm on testing

Sleepwalking commented:
Hello Tuxzz,

Thank you for your contribution. Well done.
Before merging your PR, could you please first make a few changes listed as follows?
- make AIR-QHM and dependency on LAPACK/BLAS optional in makefile
- `a->b` -> `a -> b`
- `++i` -> `i ++`
- `&a` (as reference operator)  -> `& a`
- `typename function_name(...) \n {` -> `typename function_name(...) {`
- `enum llsm_harmonic_analysis_method {...}` -> `#define LLSM_HAMETHOD_QFFT 0 ...`
- regarding code format of if...else blocks,

```
if(...) {
...
}
else {
...
}
```

->

```
if(...) {
...
} else {
...
}
```
- comment [line 583 of `llsm.c`](https://github.com/Sleepwalking/libllsm/pull/6/files#diff-aebcd37667df7ee403334b8e22e23f5cR583)
- rename `llsm_parameters.a_method` to `llsm_parameters.a_hamethod`
- `isshowprogress` -> `show_progress`
- wrap the line at column 100
- `qhm_status_free` -> `delete_qhm_status`

Grammatical errors in comments are noted in diff page of this PR.

Thanks & regards,
Kanru Hua

Sleepwalking commented on llsm.h:
```diff
@@ -109,6 +111,12 @@ typedef struct {
 */
+
+typedef enum {
+  qfft = 0, // use fft spectrum peak picking on quasi-harmonic analysis
```
fft -> FFT
on -> for

Sleepwalking commented on llsm.h:
```diff
@@ -122,6 +130,12 @@ typedef struct {
   int a_nnosband;       // number of noise bands
 
+  llsm_harmonic_analysis_method a_method; // method of quasi-harmonic analysis, qfft by default
+  char a_qhmlsmethod; // method of LS. 'Q' remains QR/LR(faster). 'S' remains SVD(better quality, very slow).
```
remains -> indicates/means/is the abbr. of
Do you mean LQ decomposition?

Sleepwalking commented on qhm.h:
```diff
@@ -0,0 +1,95 @@
+  FP_TYPE* n; // time vec. (1, nX)
+  FP_TYPE* t; // arg of cplx exp, (K, 2N + 1)
+  double complex* Ea; // matrix with cplx exp, (2K, 2N + 1)
+  double complex* Ew; // E multiply the window, (2N + 1, 2K)
```
multiply -> multiplied by

Sleepwalking commented on qhm.h:
```diff
@@ -0,0 +1,95 @@
+  double complex* Ew; // E multiply the window, (2N + 1, 2K)
+  double complex* R; // E multiply the window, (2K, 2K);
+  double complex* windowed_x; // (nX, 1)
+  double complex* lstsqb; // (2 * K, 1), contains ak and bk(outputs)
```
always put a space before left parenthesis

Sleepwalking commented on readme.md:
```diff
@@ -17,6 +17,11 @@ It is designed to be an **interpolatable**, **low-level** speech representation
 
+Requirement
+---
+A CBLAS and a LAPACKE.
```
A CBLAS-compatible and a LAPACKE-compatible library.

Sleepwalking commented on readme.md:
```diff
@@ -17,6 +17,11 @@ It is designed to be an **interpolatable**, **low-level** speech representation
+Requirement
+---
+A CBLAS and a LAPACKE.
+Tested on reference blas/lapack and OpenBLAS.
```
What do you mean by this?

Sleepwalking commented on readme.md:
```diff
@@ -35,3 +40,6 @@ Works cited
 
 * Stylianou, Yannis. "Harmonic plus noise models for speech, combined with statistical methods, for speech and speaker modification." Diss. Ecole Nationale Supérieure des Télécommunications. 1996.
 
+* Pantazis Y, Rosec O, Stylianou Y. Iterative estimation of sinusoidal signal parameters[J]. Signal Processing Letters, IEEE, 2010, 17(5): 461-464.
```
Use MLA style citation.

tuxzz committed:
some changes according to advices
1. a->b -> a -> b
2. ++i -> i ++
3. &a (as reference operator) -> & a
4. typename function_name(...) \n { -> typename function_name(...) {
5. enums -> macros
6. regarding code format of if...else blocks,
7. rename some variables and functions
8. wrap the line at column 100
9. fix grammatical errors

tuxzz committed:
make QHM-AIR and dependency on LAPACK/BLAS optional in makefile

tuxzz commented:
I've done it.
And sorry for my poor English. Thanks a lot for your correction.

Sleepwalking commented:
Hello Tuxzz,

I'm considering reverting this pull request after reconsidering the assumption underlying the adaptive (quasi) harmonic model family. In those models aperiodic component is not assumed to be independent from periodic component (rather, they are **meant to be** jointly modeled) and thus the aperiodic signal recovered by subtraction method would be underestimated, and to some extent this is still the case even when QHM parameters are converted to quasi-static form (frequency, amplitude and phase which have the same degree of freedom as QIFFT-estimated sinusoids).

tuxzz commented:
Hi Sleepwalking,

I noticed this problem in both QHM and QFFT. So I set default a_maxqhmiter as 4 rather than higher value to avoid horrible effect on noisy voice and got a good result on my test cases.
In QFFT, aperiodic component is also not fully independent from periodic component. Amplitude of Quasi-harmonic analyzed from spectrum contains aperiodic component.
So I think this is not a critical problem.

tuxzz commented:
But please NOTICE that sinusoidal frequencies produced by qhm_air may be **NOT** in order, which may be cause an error on envelope analysis step or other steps.
You need sort sinusoidal parameters by some functions such as argsort.

Sleepwalking commented:
In fact I'm working on a new version of LLSM including several major changes to the model (e.g. from sinusoidal to strictly harmonic). Later systematic evaluation of noise extraction performance will be measured and I'll get back to this issue.

Sleepwalking commented:
I think reviewing the assumptions behind QIFFT and QHM would be helpful in understanding the problem. QIFFT assumes P/AP components are **spectrally separable** while QHM **does not make the distinction** and it tries to approximate both P & AP using frequency & amplitude varying sinusoids. So theoretically I would prefer QIFFT in this very case where we really want to separate them.
However when hop size goes down sinusoidal model converges to some approximate version of QHM though the time resolution isn't that great. So none of them do a good job at separating noise, and you are right that under certain condition one does not significantly outperform the other (and as an engineer you may pick any of them as long it roughly works).
